### PR TITLE
Add a HoC for protected routes

### PR DESCRIPTION
**Pull request by maarcelh**
Added a higher order component which checks if a user is authenticated and if a user is logged in for the first time. 

If a user isnt logged in the user will be redircted to the signin page, and if a user has logged in for the first time the user is redirected to the change-password page.

Any component that needs authentication can be wrapped with this component

### CODE REVIEW

File: context/AuthContext.tsx [hunk 1]
```diff
@@ -51,7 +51,7 @@ interface AuthContextProviderProps {
 
 export const AuthContextProvider = ({children}: AuthContextProviderProps) => {
     const [user, setUser] = useState<User | null>(null);
-    const [isFirstLogin, setIsFirstLogin] = useState<boolean | undefined>(true)
+    const [isFirstLogin, setIsFirstLogin] = useState<boolean | undefined>(undefined)
```
Comment (fkoulen) [inline context/AuthContext.tsx:54]: Why can this boolean be undefined?

File: components/protected-route/ProtectedRoute.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+const WithProtectedRoute = <P extends object>(
+    WrappedComponent: FC<P>
+): FC<P> => {
+    const Wrapper = (props: P) => {
+        const router = useRouter();
+        const { user, isFirstLogin } = useAuthContext();
+        
+        if (!user) {
+            router.push("./signin");
```
Comment (fkoulen) [inline components/protected-route/ProtectedRoute.tsx:15]: Also sign in are two words so maybe rename this and the folder to sign-in

File: components/protected-route/ProtectedRoute.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        if (isFirstLogin) {
+            if(router.pathname !== "/change-password") {
+                router.push("/change-password");
+            }
+        }
+
+        return !user ? null : <WrappedComponent {...props} />;
+    };
+
+    return Wrapper;
```
Comment (fkoulen) [inline components/protected-route/ProtectedRoute.tsx:27]: Is it necessary to create a local variable for this? Perhaps inline the variable.

Comment (maarcelh) [reply components/protected-route/ProtectedRoute.tsx:27]: yes cuz hooks wont work without declaring first

File: components/protected-route/ProtectedRoute.tsx (hunk 2 above)
Comment (maarcelh) [reply components/protected-route/ProtectedRoute.tsx:15]: eehhh okay

**Comment by maarcelh** [comment]
closing this because im going for another approach to handle first login

**Pull request closed without merging**